File: sample_source_code/typescript/src/Dog.ts
import { Animal }   from './Animal';
import type { IVisitor } from './IVisitor';
import { MathUtils } from './MathUtils';

const BREED_IDEAL_WEIGHT: Record<string, number> = {
    'chihuahua':       2.5,
    'labrador':        30.0,
    'german shepherd': 35.0,
};

/** Concrete Dog with breed-specific health scoring. */
export class Dog extends Animal {
    constructor(
        name:  string,
        age:   number,
        public readonly breed: string,
        weight: number,
    ) {
        super(name, age, 'Canis lupus familiaris', weight);
    }

    sound(): string             { return 'Woof'; }
    accept(v: IVisitor): void   { v.visitDog(this); }

    /**
     * Computes a health score (0-100) based on age and weight vs breed ideal.
     * Outgoing calls to MathUtils.
     */
    computeHealthScore(): number {
        let score = 100;

        if (this._age <= 2) {
            score -= MathUtils.linearDecay(this._age, 2, 5);
        } else if (this._age <= 7) {
            score -= MathUtils.linearDecay(this._age, 7, 10);
        } else {
            score -= MathUtils.linearDecay(this._age, 15, 50);
        }

        const ideal     = BREED_IDEAL_WEIGHT[this.breed.toLowerCase()] ?? 15;
        const deviation = Math.abs(this._weight - ideal) / ideal;
        score -= deviation * 20;

        return MathUtils.clamp(Math.round(score * 100) / 100, 0, 100);
    }
}

File: sample_source_code/csharp/src/Dog.cs
namespace Embold.Sample;

/// <summary>Concrete Dog class with breed-specific health scoring.</summary>
public sealed class Dog : Animal
{
    private static readonly Dictionary<string, double> BreedIdealWeight =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["chihuahua"]       = 2.5,
            ["labrador"]        = 30.0,
            ["german shepherd"] = 35.0,
        };

    public Dog(string name, int age, string breed, double weight)
        : base(name, age, "Canis lupus familiaris", weight)
    {
        Breed = breed;
    }

    public string Breed { get; }

    public override string Sound() => "Woof";
    public override void   Accept(IVisitor visitor) => visitor.Visit(this);

    /// <summary>
    /// Computes a health score (0–100) based on age and weight vs breed ideal.
    /// Outgoing calls to <see cref="MathUtils"/>.
    /// </summary>
    public double ComputeHealthScore()
    {
        double score = 100.0;

        score -= Age switch
        {
            <= 2 => MathUtils.LinearDecay(Age, 2,  5),
            <= 7 => MathUtils.LinearDecay(Age, 7,  10),
            _    => MathUtils.LinearDecay(Age, 15, 50),
        };

        double idealWeight = BreedIdealWeight.GetValueOrDefault(Breed, 15.0);
        double deviation   = Math.Abs(Weight - idealWeight) / idealWeight;
        score -= deviation * 20;

        return MathUtils.Clamp(Math.Round(score, 2), 0, 100);
    }
}
